src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8Utility.Validation.cs

```diff
@@ -720,5 +735,22 @@ internal static unsafe partial class Utf8Utility
             scalarCountAdjustment = tempScalarCountAdjustment;
             return pInputBuffer;
         }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int AdvSimdMoveMask(Vector128<byte> value)
+        {
+            Debug.Assert(AdvSimd.Arm64.IsSupported);
+
+            // extractedBits[i] = (value[i] & 0x80) == 0x80 & (1 << i);
+            Vector128<byte> mostSignficantBitMask = s_mostSignficantBitMask;
+            Vector128<byte> mostSignificantBitIsSet = AdvSimd.CompareEqual(AdvSimd.And(value, mostSignficantBitMask), mostSignficantBitMask);
+            Vector128<byte> extractedBits = AdvSimd.And(mostSignificantBitIsSet, s_bitMask128);
+
+            // self-pairwise add until all flags have moved to the first two bytes of the vector
+            extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
+            extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
+            extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
+            return extractedBits.AsInt32().ToScalar();
```

Comment:
I realize you took this from my code, but this line contains a bug. You need to change `AsInt32()` to `AsUInt16()` to get the correct mask back.

Reply:
Also, this particular line assumes little-endian

Reply:
This comment was added before I changed the code to @kunalspathak 's suggestion, which reduced the `AddPairwise` calls. I suppose the bug you pointed out does not apply anymore?

Reply:
Correct.